keps/sig-node/127-user-namespaces/README.md

```diff
@@ -0,0 +1,393 @@
+can mitigate the impact of a container breakout vulnerability (e.g to read
+host files).
+
+#### Story 5
+
+As a cluster admin, I want to use different host UIDs/GIDs for pods running on
+the same node (whenever kernel/kube features allow it), so I can mitigate the
+impact a compromised pod can have on other pods and the node itself.
+
+### Notes/Constraints/Caveats
+
+#### UID space
+
+#### Volume support
+
+### Risks and Mitigations
+
+## Design Details
+
+### Pod.spec changes
+
+The following changes will be done to the pod.spec:
+
+- `pod.spec.hostUsers`: bool.
+If true or not present, uses the host user namespace (as today)
+If false, a new userns is created for the pod.
+By default it is set to `true`.
+
+### Phases
+
+We propose to divide the work in 3 phases. Each phase makes this work with
+either more isolation or more workloads. When no support is yet added to handle
+some workload, a clear error will be shown.
+
+PLEASE note that only phase 1 is targeted for alpha. Also note that the missing
+details (CRI changes, changes needed in container runtimes, etc.) will be added
+in a follow-up PRs.
+
+Please note the last sub-section here is a table with the summary of the changes
+proposed on each phase. That table is not updated (it is from the initial
+proposal, doesn't have all the feedback and adjustments we discussed) but can
+still be useful as a general overview.
+
+
+#### Phase 1: pods "without" volumes
+
+This phase makes pods "without" volume work with user namespaces. This is
+activated via the bool `pod.spec.HostUsers` and works only when no volume or any
+combination of the following volume types are used by the pod:
+
+ - configmap
+ - secret
+ - downwardAPI
+ - emptyDir
+ - projected
+
+This list of volumes was chosen as they can't be used to share files with other
+pods.
+
+The mapping length will be 65535, mapping the range 0-65534 to the pod. This wide
+range makes sure most workloads will work fine.
+
+The mapping will be chosen by the kubelet, using a simple algorithm to give
+different pods in this category ("without" volumes) a non-overlapping mapping.
+Giving non-overlapping mappings generates the best isolation for pods.
+
+Furthermore, the node UID space of 2^32 can hold up to 2^16 pods each with a
+mapping length of 65535 (2^16-1) top. This imposes a limit of 65k pods per node,
+but that is not an issue we will hit in practice for a long time, if ever (today
+we run 110 pods per node by default).
+
+#### Phase 2: pods with volumes
+
+This phase makes user namespaces work in pods with volumes too. This is
+activated via the bool `pod.spec.HostUsers` too and pods fall into this mode if
+some other volume type than the ones listed for phase 1 is used. IOW, when phase
+2 is implemented, pods that use volume types not supported in phase 1, fall into
+the phase 2.
+
+All pods in this mode will use _the same_ mapping, chosen by the kubelet, with a
```

Comment:
I think the mapping and size has to be specced by us, not chosen by kubelet, in order to allow a cluster with mixed node types (e.g. some containerd and some CRI-O, which might make different decisions).  But we don't have to block alpha on this.

Can you please add an UNRESOLVED section here to capture this decision point, so we can revisit before beta?

Reply:
Another UNRESOLVED: do we actually want to ship phase 2 as automatic thing?  If we do, any workloads that use it can't ever switch to whatever we do in phase 3.

Reply:
Why would we want different container runtimes to make different decisions about the mapping? We thought about this with @giuseppe in the past, but we didn't find any reason to do that (and for reasons to not do it. See https://github.com/kubernetes/enhancements/pull/3065#issuecomment-1017445796, https://github.com/kubernetes/enhancements/pull/3065#issuecomment-1017569797 (and the links I pasted there to other conversations about it) and https://github.com/kubernetes/enhancements/pull/3065#discussion_r798828286.

In any case, good idea the unresolved section, will add it. Thanks!

Reply:
regarding the container runtime picking the mappings. One more reason @giuseppe remind me we discussed: the kubelet needs to know the mappings. If not, it won't be able to give the proper ownership to configmap/secret volumes. So, if the container runtime picks the mappings, the kubelet will need to ask before creating the volumes... One more reason for the kubelet to pick the mapping :)